### PR TITLE
Add reference for Configuration to kafka cluster

### LINKED ISSUE #489: kakfa: `Cluster` resource is not able to get to the Ready state
### What happened?
I've been seeing problems with the `cluster.kafka.aws.upbound.io`, it seems that when setting `spec.forProvider.clusterName` which is required, this becomes the name of the resource in AWS, and then crossplane thinks it no longer owns it. If we set the `crossplane.io/external-name` annotation to match this, then it's unable to create the cluster, when we use external-name we get this error:

```    
Message:               observe failed: cannot run refresh: refresh failed: reading MSK Cluster (services-kafka-cluster): AccessDeniedException:
                           status code: 403, request id: 4f4bf7dc-a24e-450b-8c42-4d516be05743:
    Reason:                ReconcileError
    Status:                False
    Type:                  Synced
```

When we don't use external-name we get this error:

```
    Message:               apply failed: creating MSK Cluster (services-kafka-cluster): ConflictException: A resource with this name already exists.
{
  RespMetadata: {
    StatusCode: 409,
    RequestID: "dfce3934-d9ea-4ec8-885b-7d56b91e95a3"
  },
  InvalidParameter: "clusterName",
  Message_: "A resource with this name already exists."
}
```

Which makes no sense because crossplane has full admin in this account. I'm not sure how this can to ever get to the Ready state. 

### How can we reproduce it?
Create a Kafka cluster with the resource.

### What environment did it happen in?
* Universal Crossplane Version: 1.10.1
* Provider Version: v0.26.0
* Cloud provider or hardware configuration: AWS
* Kubernetes version (use `kubectl version`): v1.24.8-eks-ffeb93d
* Kubernetes distribution (e.g. Tectonic, GKE, OpenShift): EKS

### PR DESCRIPTION
**Pull request by mbbush**
### Description of your changes

Partially Fixes #489 

It seems that creating a kafka cluster sometimes fails, with an error related to the configuration, as described in #489. One workaround we've implemented is to always define both the `Configuration.kafka` and the `Cluster.kafka` in the same composition, and patch from the `Configuration` to the composite, then from the composite to the `Cluster` with `fromFieldPath: Required`. This is cumbersome, but it works. 

This seems like potentially a better solution, allowing the use of a selector to get essentially the same behavior, but the cost is that it requires you to specify the same selector twice. What I really wanted to do was to be able to extract both the ARN and the revision from the same object, but it seems like the output of the `ExtractValueFn` must always be a string, and `make generate` panics if I have more than one selector with the same name, so I couldn't see a way to do it. 

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

```
make uptest UPTEST_EXAMPLE_LIST='examples/kafka/cluster.yaml'
--- PASS: kuttl (2680.69s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/case (2656.51s)
PASS
13:38:27 [ OK ] running automated tests
```

### CODE REVIEW

**Comment by turkenf** [comment]
/test-examples="examples/kafka/cluster.yaml"

File: config/kafka/config.go [hunk 1]
```diff
@@ -25,6 +25,14 @@ func Configure(p *config.Provider) {
 		r.References["broker_node_group_info.security_groups"] = config.Reference{
 			Type: "github.com/upbound/provider-aws/apis/ec2/v1beta1.SecurityGroup",
 		}
+		r.References["configuration_info.arn"] = config.Reference{
+			Type:      "Configuration",
+			Extractor: common.PathARNExtractor,
+		}
+		r.References["configuration_info.revision"] = config.Reference{
+			Type:      "Configuration",
+			Extractor: "GetConfigurationRevision()",
```
Comment (turkenf) [inline config/kafka/config.go:34]:
```suggestion
			Extractor: `github.com/upbound/upjet/pkg/resource.ExtractParamPath("latest_revision",true)`,
```
We have a [function](https://github.com/upbound/upjet/blob/0154ca2d806f389c294ed475d2535a96ddb4c3c0/pkg/resource/extractor.go#L34) for such situations, could you please try the configuration above?

**Comment by turkenf** [comment]
Since it takes a long time to create and delete this resource, can you add a 2-hour timeout like [here](https://github.com/upbound/provider-aws/blob/9963d735d3fba313f3aade0843d3304e87b73ede/examples/memorydb/cluster.yaml#L6) to extend the running time of uptest?

File: config/kafka/config.go (hunk 1 above)
Comment (mbbush) [reply config/kafka/config.go:34]: Yes! I thought I remembered seeing something like that, and looked for it, but couldn't find it. This is much better.

**Comment by mbbush** [comment]
I updated the example to include the 2 hour timeout (oof) and also fixed a schema error. I don't know why the `configurationInfo` is an array, probably because of something about the way it's set up in terraform? It certainly seems like it would make more sense as an object in yaml, but I guess that would be a breaking schema change to a v1beta1 resource, which probably isn't worth it. 

For some reason when I try to run the e2e test locally I get this error: 

```
    case.go:363: DeliveryStream.firehose.aws.upbound.io "test-stream" is invalid: spec: Invalid value: "object": no such key: initProvider evaluating rule: name is a required parameter
```

Any idea why that might be?

**Comment by turkenf** [comment]
> Any idea why that might be?

Can you use the example [here](https://github.com/upbound/provider-aws/blob/c58e73e035dfe1e05f074261a4685e7ec88505d8/examples/kafka/cluster.yaml#L76) for the `DeliveryStream.firehose` resource? The example in `Cluster.kafka` resource has not been updated and has deprecated parameters.

**Comment by mbbush** [comment]
ok, that's fixed. Now the problem is that the `Cluster.kafka` is having a reconcile error:
```
    Last Transition Time:  2023-09-14T21:20:46Z
    Message:               cannot resolve references: mg.Spec.ForProvider.ConfigurationInfo[i3].Revision: referenced field was empty (referenced resource may not yet be ready)
    Reason:                ReconcileError
    Status:                False
    Type:                  Synced
```

even though the `Configuration` is Ready:
```

kubectl get -o yaml configuration.kafka
apiVersion: v1
items:
- apiVersion: kafka.aws.upbound.io/v1beta1
  kind: Configuration
  metadata:
    annotations:
      crossplane.io/external-create-succeeded: "2023-09-14T21:18:28Z"
      crossplane.io/external-name: arn:aws:kafka:us-west-1:226209437353:configuration/example/c8c1a33a-54f0-44ef-8a5c-a99ce0f33319-2
      upjet.crossplane.io/provider-meta: "null"
      upjet.upbound.io/test: "true"
    creationTimestamp: "2023-09-14T21:18:28Z"
    finalizers:
    - finalizer.managedresource.crossplane.io
    generation: 2
    labels:
      testing.upbound.io/example-name: example
    name: example
    resourceVersion: "153241"
    uid: aa1f90fe-51b0-46d3-b9d1-51c91490bbb7
  spec:
    deletionPolicy: Delete
    forProvider:
      kafkaVersions:
      - 2.6.0
      name: example
      region: us-west-1
      serverProperties: |
        auto.create.topics.enable = true
        delete.topic.enable = true
    initProvider: {}
    managementPolicies:
    - '*'
    providerConfigRef:
      name: default
  status:
    atProvider:
      arn: arn:aws:kafka:us-west-1:226209437353:configuration/example/c8c1a33a-54f0-44ef-8a5c-a99ce0f33319-2
      description: ""
      id: arn:aws:kafka:us-west-1:226209437353:configuration/example/c8c1a33a-54f0-44ef-8a5c-a99ce0f33319-2
      kafkaVersions:
      - 2.6.0
      latestRevision: 1
      name: example
      serverProperties: |
        auto.create.topics.enable = true
        delete.topic.enable = true
    conditions:
    - lastTransitionTime: "2023-09-14T21:18:34Z"
      reason: Available
      status: "True"
      type: Ready
    - lastTransitionTime: "2023-09-14T21:18:29Z"
      reason: ReconcileSuccess
      status: "True"
      type: Synced
    - lastTransitionTime: "2023-09-14T21:18:31Z"
      reason: Success
      status: "True"
      type: LastAsyncOperation
    - lastTransitionTime: "2023-09-14T21:18:31Z"
      reason: Finished
      status: "True"
      type: AsyncOperation
    - lastTransitionTime: "2023-09-14T21:18:45Z"
      reason: UpToDate
      status: "True"
      type: Test
kind: List
metadata:
  resourceVersion: ""
```

**Comment by mbbush** [comment]
Oh, I bet the problem is that `status.atProvider.latestRevision` is a number, not a string.

**Comment by turkenf** [comment]
Then we can consider canceling the reference definition for the revision parameter and passing the value as `revision: 1`. What do you say?

**Comment by mbbush** [comment]
That will make the test pass. I already tried that locally and it worked. It does make the reference substantially less useful, but I suppose it's still better than nothing? 

How deeply baked into the code for resolving references is the requirement that the resolved value must be a string? What I'd really like to be able to do is use a Resolver that returns an object, which I can customize. Something like (pseudocode)

```
def resolveConfiguration(mr: xp.ManagedResource) = {
  return {
    "arn": mr.status.atProvider.arn,
    "revision": mr.status.atProvider.latestRevision
   }
}
```

This seems useful in general as a flexible (ideally user-definable) way to define arbitrary dependencies between managed resources. Are you aware of any design discussions around possibly supporting this, or on why it would be prohibitively difficult?

**Comment by turkenf** [comment]
/test-examples="examples/kafka/cluster.yaml"

**Comment by turkenf** [comment]
For checking uptest logs: https://github.com/upbound/provider-gcp/pull/391#issuecomment-1725071475

**Comment by mbbush** [comment]
As far as I can tell the test failed because one or more of the subnets didn't become ready. 

```
2023-09-20T08:29:51Z	DEBUG	provider-aws	apply async ended	{"workspace": "/tmp/88afe8fa-8857-45c1-883e-737f847b19d4", "out": "{\"@level\":\"info\",\"@message\":\"Terraform 1.5.5\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:50.081714Z\",\"terraform\":\"1.5.5\",\"type\":\"version\",\"ui\":\"1.1\"}\n{\"@level\":\"info\",\"@message\":\"aws_subnet.subnet-az2: Plan to create\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:51.314176Z\",\"change\":{\"resource\":{\"addr\":\"aws_subnet.subnet-az2\",\"module\":\"\",\"resource\":\"aws_subnet.subnet-az2\",\"implied_provider\":\"aws\",\"resource_type\":\"aws_subnet\",\"resource_name\":\"subnet-az2\",\"resource_key\":null},\"action\":\"create\"},\"type\":\"planned_change\"}\n{\"@level\":\"info\",\"@message\":\"Plan: 1 to add, 0 to change, 0 to destroy.\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:51.314247Z\",\"changes\":{\"add\":1,\"change\":0,\"import\":0,\"remove\":0,\"operation\":\"plan\"},\"type\":\"change_summary\"}\n{\"@level\":\"info\",\"@message\":\"aws_subnet.subnet-az2: Creating...\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:51.800935Z\",\"hook\":{\"resource\":{\"addr\":\"aws_subnet.subnet-az2\",\"module\":\"\",\"resource\":\"aws_subnet.subnet-az2\",\"implied_provider\":\"aws\",\"resource_type\":\"aws_subnet\",\"resource_name\":\"subnet-az2\",\"resource_key\":null},\"action\":\"create\"},\"type\":\"apply_start\"}\n{\"@level\":\"info\",\"@message\":\"aws_subnet.subnet-az2: Creation errored after 0s\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:51.946235Z\",\"hook\":{\"resource\":{\"addr\":\"aws_subnet.subnet-az2\",\"module\":\"\",\"resource\":\"aws_subnet.subnet-az2\",\"implied_provider\":\"aws\",\"resource_type\":\"aws_subnet\",\"resource_name\":\"subnet-az2\",\"resource_key\":null},\"action\":\"create\",\"elapsed_seconds\":0},\"type\":\"apply_errored\"}\n{\"@level\":\"error\",\"@message\":\"Error: creating EC2 Subnet: InvalidParameterValue: Value (REDACTEDc) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: REDACTEDa, REDACTEDb.\\n\\tstatus code: 400, request id: c748bd40-b8df-4902-89ee-8ac002585416\",\"@module\":\"terraform.ui\",\"@timestamp\":\"2023-09-20T08:29:51.947437Z\",\"diagnostic\":{\"severity\":\"error\",\"summary\":\"creating EC2 Subnet: InvalidParameterValue: Value (REDACTEDc) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: REDACTEDa, REDACTEDb.\\n\\tstatus code: 400, request id: c748bd40-b8df-4902-89ee-8ac002585416\",\"detail\":\"\",\"address\":\"aws_subnet.subnet-az2\",\"range\":{\"filename\":\"main.tf.json\",\"start\":{\"line\":1,\"column\":443,\"byte\":442},\"end\":{\"line\":1,\"column\":444,\"byte\":443}},\"snippet\":{\"context\":\"resource.aws_subnet.subnet-az2\",\"code\":\"{\\\"provider\\\":{\\\"aws\\\":{\\\"access_key\\\":\\\"REDACTED\\\",\\\"region\\\":\\\"REDACTED\\\",\\\"secret_key\\\":\\\"REDACTED\\\",\\\"token\\\":\\\"\\\"}},\\\"resource\\\":{\\\"aws_subnet\\\":{\\\"subnet-az2\\\":{\\\"availability_zone\\\":\\\"REDACTEDc\\\",\\\"cidr_block\\\":\\\"192.168.1.0/24\\\",\\\"lifecycle\\\":{\\\"prevent_destroy\\\":true},\\\"tags\\\":{\\\"crossplane-kind\\\":\\\"subnet.ec2.aws.upbound.io\\\",\\\"crossplane-name\\\":\\\"subnet-az2\\\",\\\"crossplane-providerconfig\\\":\\\"default\\\"},\\\"vpc_id\\\":\\\"vpc-071bb13dcb266c3d4\\\"}}},\\\"terraform\\\":{\\\"required_providers\\\":{\\\"aws\\\":{\\\"source\\\":\\\"hashicorp/aws\\\",\\\"version\\\":\\\"4.67.0\\\"}}}}\",\"start_line\":1,\"highlight_start_offset\":442,\"highlight_end_offset\":443,\"values\":[]}},\"type\":\"diagnostic\"}\n"}
```

Maybe for some reason your aws account is only allowed to use the `us-west-1a` and `us-west-1b` AZs, and not `us-west-1c`? Or maybe AWS was having some problem in `us-west-1c` yesterday?

I ran the test twice locally on my laptop. The first time, after it had requested creation of the cluster, one of the aws api calls failed with a network error, probably because I was switching between wifi and ethernet, which caused the provider to `terraform taint` the kafka cluster. It was created in AWS, but never became ready in crossplane until I logged into the provider pod and manually `terraform untaint`ed it. Then it worked fine. 

I reran the test a second time and left my computer plugged in to ethernet the whole time. The test passed without intervention, and those are the results I posted above.

**Comment by turkenf** [comment]
> Maybe for some reason your aws account is only allowed to use the us-west-1a and us-west-1b AZs, and not us-west-1c? Or maybe AWS was having some problem in us-west-1c yesterday?

Yes, it is, if this run also fails due to this, you can use `us-west-1a` and `us-west-1b`, I will trigger now.

**Comment by turkenf** [comment]
/test-examples="examples/kafka/cluster.yaml"

**Comment by mbbush** [comment]
I moved the whole example to `us-east-2` and switched to availability zones a and b just before you triggered the test, so :crossed_fingers: it will pass this time.

**Review by turkenf — APPROVED** [review]
Thanks so much for your work in this PR @mbbush, LGTM.

**Pull request merged**